cmd/pipecd/ops.go

```diff
@@ -79,6 +80,35 @@ func (s *ops) run(ctx context.Context, t cli.Telemetry) error {
 		return err
 	}
 
+	// Prepare sql database.
+	if cfg.Datastore.Type == model.DataStoreMySQL {
+		mysqlEnsurer := mysqlensurer.NewMySQLEnsurer(
+			cfg.Datastore.MySQLConfig.URL,
+			cfg.Datastore.MySQLConfig.Database,
+			cfg.Datastore.MySQLConfig.UsernameFile,
+			cfg.Datastore.MySQLConfig.PasswordFile,
+			t.Logger,
+		)
+		defer func() {
+			// Close connection held by the client.
+			if err := mysqlEnsurer.Close(); err != nil {
+				t.Logger.Error("failed to close database ensurer connection", zap.Error(err))
+			}
+		}()
```

Comment:
I meant wrapping the whole of this part into a function to use defer.
Using `defer` like this will keep the connection until `ops` is terminated. We should close that DB connection after finishing those tasks.

Reply:
Right. I agree with that. Making this part an isolated function seems good.

Reply:
> Using defer like this will keep the connection until ops is terminated.
Agree, the scope of `defer` enclose to function, not the lexical enclosure, that why I make `Close` directly at the first version.

>  wrapping the whole of this part into a function to use defer
Just to clarify, do you mean making a kind of `ensureSQLDatabase` function and defer close() in that function, then
```
	// Prepare sql database.
	if cfg.Datastore.Type == model.DataStoreMySQL {
		if err := ensureSQLDatabase(); err != nil {
			t.Logger.Error("failed to create required indexes on sql database", zap.Error(err))
			return err
		}
	}
```
in `run()` 😂

Reply:
Yes, you can break it to be an independent function or just a function inside a function. It's up to you. 👍

Reply:
addressed on https://github.com/pipe-cd/pipe/pull/1699/commits/d1620dbde6160e4716b9133e388231961ade9700 🙏